include/simdjson/inline/document_stream.h

```diff
@@ -5,10 +5,70 @@
 #include <algorithm>
 #include <limits>
 #include <stdexcept>
-
 namespace simdjson {
 namespace dom {
 
+#ifdef SIMDJSON_THREADS_ENABLED
+namespace {
+  // Putting those inside the stage1_worker class may create problems with move constructors
+  // since a mutex is not supposed to be moved.
+  std::mutex m{};
+  std::condition_variable cv{};
+}
+inline void stage1_worker::finish() {
+  std::unique_lock<std::mutex> lock(m);
+  cv.wait(lock, [this]{return has_work == false;});
+}
+
+inline stage1_worker::~stage1_worker() {
+  stop_thread();
+}
+
+inline void stage1_worker::start_thread() {
+  thread = std::thread([this]{
+      while(can_work) {
+        std::unique_lock<std::mutex> thread_lock(m);
+        cv.wait(thread_lock, [this]{return has_work || !can_work;});
+        if(!can_work) {
+          break;
+        }
+        this->owner->stage1_thread_error = this->owner->run_stage1(*this->stage1_thread_parser,
+              this->_next_batch_start);
+        this->has_work = false;
+        thread_lock.unlock();
+        cv.notify_one(); // will notify "finish"
+      }
+    }
+  );
+}
+
+
+inline void stage1_worker::stop_thread() {
+  std::unique_lock<std::mutex> lock(m);
+  // We have to make sure that all locks can be released.
+  can_work = false;
+  has_work = false;
+  lock.unlock();
+  cv.notify_all();
+  if(thread.joinable()) {
+    thread.join();
+  }
+}
+
+inline void stage1_worker::run(document_stream * ds, dom::parser * stage1, size_t next_batch_start) {
+    if(!thread.joinable()) {
```

Comment:
I can't think of a good reason this condition would get triggered more than once, but it worries me and threads are very good at finding ways to do exactly that sort of thing. I think I'd be more comfortable if document_stream called start_thread() directly, or else maybe have a `bool first_time` parameter on run() and set that to `true` in document_stream::start(). But I'm sure it'll be fine, probably, maybe :)

Reply:
Right. Possible the "joinable" call could be expensive. Let me fix that.

Reply:
I worry more about what happens when we call start_thread() multiple times--can we end up with multiple running threads? Do we accidentally kill a thread we need because of a race we didn't think of?

The expense of the joinable() call really shouldn't be a huge factor; it only happens once per batch, and batches are allowed to be big, or "big" :)

Reply:
> I worry more about what happens when we call start_thread() multiple times

Ok. So we can lock this function so that if there are multiple calls, they won't happen at the same time. Furthermore, we can check, in the function, in the locked function, whether the thread is joinable. This should make it flat out impossible to create the thread twice.

Note that in the code, we have just one place where we start the thread now... it is called from the document_stream, within the "start" function.

Reply:
Yeah, having only one place it's possible to start the thread resolves the problem for me :)